src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs

```diff
@@ -69,49 +86,64 @@ public ProjectImportsCollector(string logFilePath, bool createFile, string sourc
+        public void AddFile(string? filePath)
+        {
+            AddFileHelper(filePath, AddFileCore);
+        }
+
+        public void AddFileFromMemory(
+            string? filePath,
+            string data,
+            DateTimeOffset? entryCreationStamp = null,
+            bool makePathAbsolute = true)
+        {
+            AddFileHelper(filePath, path =>
+                AddFileFromMemoryCore(path, data, makePathAbsolute, entryCreationStamp));
+        }
+
+        public void AddFileFromMemory(
+            string? filePath,
+            Stream data,
+            DateTimeOffset? entryCreationStamp = null,
+            bool makePathAbsolute = true)
+        {
+            AddFileHelper(filePath, path => AddFileFromMemoryCore(path, data, makePathAbsolute, entryCreationStamp));
+        }
+
+        private void AddFileHelper(
+            string? filePath,
+            Action<string> addFileWorker)
         {
             if (filePath != null && _fileStream != null)
             {
                 lock (_fileStream)
                 {
-                    // enqueue the task to add a file and return quickly
-                    // to avoid holding up the current thread
-                    _currentTask = _currentTask.ContinueWith(t =>
+                    if (_runOnBackground)
+                    {
+                        // enqueue the task to add a file and return quickly
+                        // to avoid holding up the current thread
+                        _currentTask = _currentTask.ContinueWith(
+                            t => { TryAddFile(); },
+                            TaskScheduler.Default);
+                    }
+                    else
                     {
-                        try
-                        {
-                            AddFileCore(filePath);
-                        }
-                        catch
-                        {
-                        }
-                    }, TaskScheduler.Default);
+                        TryAddFile();
+                    }
                 }
             }
-        }
 
-        public void AddFileFromMemory(string filePath, string data)
-        {
-            if (filePath != null && data != null && _fileStream != null)
+            void TryAddFile()
             {
-                lock (_fileStream)
+                try
                 {
-                    // enqueue the task to add a file and return quickly
-                    // to avoid holding up the current thread
-                    _currentTask = _currentTask.ContinueWith(t =>
-                    {
-                        try
-                        {
-                            AddFileFromMemoryCore(filePath, data);
-                        }
-                        catch
-                        {
-                        }
-                    }, TaskScheduler.Default);
+                    addFileWorker(filePath);
                 }
+                catch
```

Comment:
>catch [](http://example.com/codeflow?start=16&length=5)

Is there a specific set of exceptions we could be handling? Is there some logging, etc., that could/should be done here?

Reply:
This is just wrapping pre-existing code:

https://github.com/dotnet/msbuild/blob/f4b2350acc97e8e29e0d8498a082b9f91115f0b6/src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs#L84-L90

I intentionally chose not to alter it here (adding files to binlog is an action that should not influence success of build or log creation. We might want to add info message into the log about inability of embedding particular file - but that is currently out of scope of this change)

Reply:
Tracked as https://github.com/dotnet/msbuild/issues/9450